### PR TITLE
ImageNet-Patch dataset

### PR DESCRIPTION
**Pull request by maurapintor**
# Task Description
A dataset for benchmarking machine learning robustness against adversarial patches, based on the ImageNet dataset and adversarial patch attacks.

image

## Dataset Creation
Images are taken from the ImageNet dataset. The ten optimized patches proposed in the paper are then applied to such images to create the perturbed data.
The patch optimization follows [1].

[1] https://arxiv.org/abs/1712.09665

image

## Evaluation Metrics

- Robust accuracy: correct classification of the images with the applied patch.

## Expected Insights/Relevance

This dataset was designed to measure the robustness of the models to the application of adversarial patches. Adding this test can give further insights on robustness of the model as a general property, beyond the evaluation that is usually performed against one specific perturbation bound (e.g., $L_\infty$ norm with maximum size of 8/255)

## Access

We release the dataset on Zenodo (https://zenodo.org/record/6568778), as well as the code to generate patches and to apply them dynamically to custom images (https://github.com/pralab/ImageNet-Patch).

## Data

The dataset is hosted on Zenodo: https://zenodo.org/record/6568778

## License

We released the data with the [Creative Commons Attribution 4.0 International](https://creativecommons.org/licenses/by/4.0/legalcode).

### CODE REVIEW

File: examples/imagenet_patch.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
+import dataclasses
```
Comment (stes) [inline examples/imagenet_patch.py:1]: A docstring here would be great.

File: examples/imagenet_patch.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+from shifthappens.models import base as sh_models
+from shifthappens.models.base import PredictionTargets
+from shifthappens.tasks.base import abstract_variable, parameter, variable
+from shifthappens.tasks.base import Task
+from shifthappens.tasks.metrics import Metric
+from shifthappens.tasks.task_result import TaskResult
+
+
+@dataclasses.dataclass
+class ImageNetPatchTarget(Task):
```
Comment (stes) [inline examples/imagenet_patch.py:27]: Also here, a docstring would be nice

File: examples/imagenet_patch.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+# patch corruptions
+
+
+@sh_benchmark.register_task(
+    name="ImageNet-Patch (cellular telephone)",
+    relative_data_folder="imagenet_patch",
+    standalone=False,
+)
+@dataclasses.dataclass
+class ImageNetPatchCellularTelephone(ImageNetPatchTarget):
```
Comment (stes) [inline examples/imagenet_patch.py:97]: Short docstrings for each class (as they appear in the documentation) would be nice.

**Review by stes — CHANGES_REQUESTED** [review]
Hi @maurapintor, thanks a lot for your contribution. We are working towards the first release of the benchmark and for this, released our [contribution guide](https://github.com/shift-happens-benchmark/icml-2022/blob/main/CONTRIBUTING.rst) with the aim of unifying the implementation of all contributions.

Please have a look, the required changes should be relatively small and are mostly concerned with placing the files within the package, and adding documentation.

From reviewing your PR, the main required changes are:

- [ ] Place the files into their own package inside the `tasks/` folder
- [ ] Adding a docstring to your package (this will be rendered in the documentation)
- [ ] Add a README and optionally citation info
- [ ] We are also looking into the best ways to perform unit or integration tests and would be happy to get some feedback on how this would be most effective for you benchmark. Would it be possible to add a "dry-run/demo" option to the benchmark, and e.g. run it only on a small subset of the data? No immediate action/update is required right now, we are first gathering opinions from all authors on this.

Thanks a lot for your contribution, and please ping me if you have any questions. If you have technical questions, also feel free to ping @kotekjedi who is happy to help with implementation details.

**Comment by stes** [comment]
Hi!

This is an automated report for the documentation coverage of your PR. To provide good user experience for the benchmark, we aim to cover most functions with doc strings, using the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

We would greatly appreciate your help in assuring that docstring coverage is sufficiently high for your contribution.

The current coverage is:
| Name                                 |   Total |   Miss |   Cover |   Cover% |
|--------------------------------------|---------|--------|---------|----------|
| examples/imagenet_patch.py           |      15 |     14 |       1 |       7% |
| shifthappens/models/torchvision.py   |       4 |      0 |       4 |     100% |
|--------------------------------------|---------|--------|---------|----------|
| TOTAL                                |      19 |     14 |       5 |    26.3% |
RESULT: FAILED (minimum: 80.0%, actual: 26.3%)

You can reproduce this result by running

```
interrogate -v --ignore-nested-functions --ignore-private --ignore-semiprivate --ignore-init-method
```

on the files you added to your PR.

File: shifthappens/tasks/imagenet_patch/__main__.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+        test_transform = tv_transforms.Compose(
+            [
+                tv_transforms.ToTensor(),
+                tv_transforms.Lambda(lambda x: x.permute(1, 2, 0)),
+            ]
+        )
+
+        dataset_folder = os.path.join(dataset_folder, str(self.target))
+        self.ch_dataset = tv_datasets.ImageFolder(
```
[the remark this answers is not in the document]
Comment (maurapintor) [reply shifthappens/tasks/imagenet_patch/__main__.py:68]: Yes, it makes sense. Thanks!

File: shifthappens/tasks/imagenet_patch/imagenet_patch.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    def _evaluate(self, model: sh_models.Model) -> TaskResult:
+        dataloader = self._prepare_dataloader()
+        all_predicted_labels_list = []
+        for predictions in model.predict(
+            dataloader, PredictionTargets(class_labels=True)
+        ):
+            all_predicted_labels_list.append(predictions.class_labels)
+        all_preds = np.concatenate(all_predicted_labels_list, 0)
+
+        accuracy = all_preds == np.array(self.ch_dataset.targets)[: len(all_preds)]
```
Comment (kotekjedi) [inline shifthappens/tasks/imagenet_patch/imagenet_patch.py:89]:
```suggestion
        accuracy = (all_preds == np.array(self.ch_dataset.targets)[: len(all_preds)]).mean()
```

**Comment by kotekjedi** [comment]
Hello Maura!
Could you please check "Allow edits from maintainers" so it would be possible for us to add minor fixes to PR before merging?

**Comment by maurapintor** [comment]
> Hello Maura!
> Could you please check "Allow edits from maintainers" so it would be possible for us to add minor fixes to PR before merging?

Thank you @kotekjedi!
I checked the box, now maintainers should be able to apply edits.

**Pull request merged**